src/steps/os/macos.rs

```diff
@@ -93,3 +94,121 @@ pub fn run_sparkle(ctx: &ExecutionContext) -> Result<()> {
+    let xcodes = require("xcodes")?;
+    print_separator("Xcodes");
+
+    let should_ask = !(ctx.config().yes(Step::Xcodes) || ctx.config().dry_run());
+
+    let releases = ctx
+        .run_type()
+        .execute(&xcodes)
+        .args(["update"])
+        .output_checked_utf8()?
+        .stdout;
+
+    let releases_installed: Vec<String> = releases
+        .lines()
+        .filter(|r| r.contains("(Installed)"))
+        .map(String::from)
+        .collect();
+
+    if releases_installed.is_empty() {
+        println!("No Xcode releases installed.");
+        return Ok(());
+    }
+
+    let (installed_gm, installed_beta, installed_regular) =
+        releases_installed
+            .iter()
+            .fold((false, false, false), |(gm, beta, regular), release| {
+                (
+                    gm || release.contains("GM"),
+                    beta || release.contains("Beta"),
+                    regular || !(release.contains("GM") || release.contains("Beta")),
+                )
+            });
+
+    let releases_gm = releases
+        .lines()
+        .filter(|&r| r.matches("GM").count() > 0)
+        .map(String::from)
+        .collect();
+    let releases_beta = releases
+        .lines()
+        .filter(|&r| r.matches("Beta").count() > 0)
+        .map(String::from)
+        .collect();
+    let releases_regular = releases
+        .lines()
+        .filter(|&r| r.matches("GM").count() == 0 && r.matches("Beta").count() == 0)
+        .map(String::from)
+        .collect();
+
+    if installed_gm {
+        let _ = process_xcodes_releases(releases_gm, should_ask, ctx);
+    }
+    if installed_beta {
+        let _ = process_xcodes_releases(releases_beta, should_ask, ctx);
+    }
+    if installed_regular {
+        let _ = process_xcodes_releases(releases_regular, should_ask, ctx);
+    }
+
+    let releases_new = ctx
+        .run_type()
+        .execute(&xcodes)
+        .args(["list"])
+        .output_checked_utf8()?
+        .stdout;
+
+    let releases_new_installed: HashSet<_> = releases_new
+        .lines()
+        .filter(|release| release.contains("(Installed)"))
+        .collect();
+
+    if should_ask && releases_new_installed.len() == 2 {
+        let answer_uninstall = prompt_yesno("Would you like to move the former Xcode release to the trash?")?;
+        if answer_uninstall {
+            let _ = ctx
+                .run_type()
+                .execute(&xcodes)
+                .args([
+                    "uninstall",
+                    releases_new_installed.iter().next().cloned().unwrap_or_default(),
+                ])
+                .status_checked();
+        }
+    }
+
+    Ok(())
+}
+
+pub fn process_xcodes_releases(releases_filtered: Vec<String>, should_ask: bool, ctx: &ExecutionContext) -> Result<()> {
+    let xcodes = require("xcodes")?;
+
+    if releases_filtered
+        .last()
+        .map(|s| !s.contains("(Installed)"))
+        .unwrap_or(true)
```

Comment:
If `releases_filtered` is empty, then `.last()` will return `None`, and we treat `None` as `true`, then we will run:

```sh
$ xcodes install <an empty string>
```

I believe this won't happen in reality as there are so many Xcode releases, but I believe we should check if `releases_filtered` is empty or not before actually involving this function, if it is empty, we should directly return.

Reply:
This is checked in `update_xcodes`, but I do agree that it is a good idea to directly return in the scenario that it is empty.